server/src/main/java/org/opensearch/search/aggregations/bucket/range/RangeAggregator.java

```diff
@@ -333,52 +356,106 @@ public void collect(int doc, long bucket) throws IOException {
-                int lo = lowBound, hi = ranges.length - 1; // all candidates are between these indexes
-                int mid = (lo + hi) >>> 1;
-                while (lo <= hi) {
-                    if (value < ranges[mid].from) {
-                        hi = mid - 1;
-                    } else if (value >= maxTo[mid]) {
-                        lo = mid + 1;
-                    } else {
-                        break;
+                MatchedRange range = new MatchedRange(ranges, lowBound, value);
+                for (int i = range.startLo; i <= range.endHi; ++i) {
+                    if (ranges[i].matches(value)) {
+                        collectBucket(sub, doc, subBucketOrdinal(owningBucketOrdinal, i));
                     }
-                    mid = (lo + hi) >>> 1;
                 }
-                if (lo > hi) return lo; // no potential candidate
-
-                // binary search the lower bound
-                int startLo = lo, startHi = mid;
-                while (startLo <= startHi) {
-                    final int startMid = (startLo + startHi) >>> 1;
-                    if (value >= maxTo[startMid]) {
-                        startLo = startMid + 1;
-                    } else {
-                        startHi = startMid - 1;
-                    }
+                return range.endHi + 1;
+            }
+        };
+    }
+
+    private void preComputeWithStarTree(LeafReaderContext ctx, CompositeIndexFieldInfo starTree) throws IOException {
+        StarTreeBucketCollector starTreeBucketCollector = getStarTreeBucketCollector(ctx, starTree, null);
+        FixedBitSet matchingDocsBitSet = starTreeBucketCollector.getMatchingDocsBitSet();
+
+        int numBits = matchingDocsBitSet.length();
+
+        if (numBits > 0) {
+            for (int bit = matchingDocsBitSet.nextSetBit(0); bit != DocIdSetIterator.NO_MORE_DOCS; bit = (bit + 1 < numBits)
+                ? matchingDocsBitSet.nextSetBit(bit + 1)
+                : DocIdSetIterator.NO_MORE_DOCS) {
+                starTreeBucketCollector.collectStarTreeEntry(bit, 0);
+            }
+        }
+    }
+
+    @Override
+    public StarTreeBucketCollector getStarTreeBucketCollector(
+        LeafReaderContext ctx,
+        CompositeIndexFieldInfo starTree,
+        StarTreeBucketCollector parentCollector
+    ) throws IOException {
+        assert parentCollector == null;
+        StarTreeValues starTreeValues = StarTreeQueryHelper.getStarTreeValues(ctx, starTree);
+        return new StarTreeBucketCollector(
+            starTreeValues,
+            StarTreeTraversalUtil.getStarTreeResult(
+                starTreeValues,
+                StarTreeQueryHelper.mergeDimensionFilterIfNotExists(
+                    context.getQueryShardContext().getStarTreeQueryContext().getBaseQueryStarTreeFilter(),
+                    fieldName,
+                    List.of(DimensionFilter.MATCH_ALL_DEFAULT)
```

Comment:
There is a possibility to optimize this further by adding a range filter of min and max of all the ranges. 

Ideal case is to add filters only for the list of ranges which are passed to the aggs.

But on the flip side, having match all filter does simplify the implementation of nested aggs and boolean query etc , since we just have to worry about the "query" part and all aggregations use "match all" filter. Later we can revisit this for optimization.

@msfroh @expani what do you folks think ?

Reply:
I think we can certainly revisit this as a TODO later since I want to keep things expandable now for nested aggregations and boolean query for now. We may have to evaluate with benchmarks if this optimization results in significant gains or not paired with other queries.

For now, I can add up a TODO `Evaluate optimizing StarTree traversal filter with specific ranges instead of MATCH_ALL`, what do you think?

Reply:
Yes I think it works for now